sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -74,9 +81,13 @@ class ExpressionSet protected(
   }
 
   override def -(elem: Expression): ExpressionSet = {
-    val newBaseSet = baseSet.clone().filterNot(_ == elem.canonicalized)
-    val newOriginals = originals.clone().filterNot(_.canonicalized == elem.canonicalized)
-    new ExpressionSet(newBaseSet, newOriginals)
+    if (elem.deterministic) {
+      val newBaseSet = baseSet.clone().filterNot(_ == elem.canonicalized)
+      val newOriginals = originals.clone().filterNot(_.canonicalized == elem.canonicalized)
+      new ExpressionSet(newBaseSet, newOriginals)
+    } else {
+      new ExpressionSet(baseSet.clone(), originals.clone())
```

Comment:
we need to drop non-deterministic `elem` from `originals`?

Reply:
No, it is not dropping it. Any non-deterministic `elem` is not considered as contained in the Set.

Reply:
`Set`? It seems `originals` is `ArrayBuffer`?

Reply:
Sorry, for `Set` I mean the whole `ExpressionSet`.
Please check the logic in `CombineFilters` and my new test case, you will understand me immediately.

Reply:
There is no need to drop, since the non-deterministic `elem` is not in `originals`.

Reply:
If so, why you clone here?

Reply:
I am trying to be consistent with the behavior of original implementation.
I think I had better override `--` for efficiency